File: 04_study/Hyunchul/First_Missing_Positive.cpp
class Solution {
public:
	int firstMissingPositive(int A[], int n) {
		int result = (n + 1);

		int positiveAry[n];
		for (int i = 0; i < n; i++) {
			positiveAry[i] = 0;
		}

		for (int i = 0; i < n; i++) {
			if (A[i] > 0 && A[i] <= n) {
				positiveAry[A[i] - 1] = A[i];
			}
		}

		if (positiveAry[0] == 0) {
			result = 1;
			return result;
		}

		for (int i = 0; i < n; i++) {

			if (positiveAry[i] != (i + 1)) {
				result = (i + 1);
				break;
			}
		}

		return result;
	}
};

File: 04_study/Gyeongwook/firstMissingPositive.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def firstMissingPositive(self, A):
        arrayLen = len(A)
        
        for idx in range(0, arrayLen) :
            if A[idx] < 1 :
                continue

            temp = A[idx]
            A[idx] = -1

            if temp <= arrayLen :
                self.sort(A, temp)
                            
        for idx in range(0, arrayLen) :
            if A[idx] < 1 :
                return idx + 1
                
        return arrayLen + 1

    def sort(self, A, startIdx):
        currentIdx = startIdx

        while True :
            if A[currentIdx - 1] != currentIdx and A[currentIdx - 1] > 0 :
                temp = A[currentIdx - 1]
                A[currentIdx - 1] = currentIdx
                currentIdx = temp
            else :
                A[currentIdx - 1] = currentIdx
                return
